File: backend/flaskr/schemas/product/category_schema.py
from flaskr import ma
from flaskr.models.product.category import Category
from flaskr.schemas import fields, Length, validates, ValidationError


class CategorySchema(ma.SQLAlchemyAutoSchema):
    name = fields.Str(required=True, validate=Length(min=2, max=30))
    image = fields.Str(required=True)

    created_at = fields.DateTime(dump_only=True, data_key="createdAt")
    updated_at = fields.DateTime(dump_only=True, data_key="updatedAt")

    @validates("name")
    def validate_name(self, name):
        if Category.query.filter(Category.name == name).first():
            raise ValidationError("Name already exists! Please try a different Name")

    class Meta:
        model = Category
        load_instance = True

File: frontend/src/features/categories/schemas/categorySchema.ts
import { z } from "zod";

export const categorySchema = z.object({
    id: z.number(),
    name: z.string().min(1, "Category Name is required"),
    image: z.string().min(1, "Category Image is required"),
});

export type Category = z.infer<typeof categorySchema>;
